File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        // if (root.left == null && root.right == null)
            // return true;
        return isBST(root.left, root.val, null) && isBST(root.right, null, root.val);
    }
    
    private boolean isBST(TreeNode node, Integer max, Integer min) {
        if (node == null)
            return true;
        
        if (max != null && node.val >= max)
            return false;
        
        if (min != null && node.val <= min)
            return false;
        
        return isBST(node.left, node.val, min) && isBST(node.right, max, node.val);
    }
}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def isBst(root, high, low):
            if root == None:
                return True
            if low != None and root.val <= low:
                return False
            if high != None and root.val >= high:
                return False
            return isBst(root.left, root.val, low) and isBst(root.right, high, root.val)
        
        return isBst(root, None, None)